yugabyte/src/yugabyte/ysql/client.clj

```diff
@@ -0,0 +1,205 @@
+            [jepsen.util :as util]
+            [jepsen.control.net :as cn]
+            [jepsen.reconnect :as rc]
+            [dom-top.core :as dt]
+            [wall.hack :as wh]
+            [slingshot.slingshot :refer [try+ throw+]]))
+
+(def timeout-delay "Default timeout for operations in ms" 20000)
+(def max-timeout "Longest timeout, in ms" 40000)
+
+(def isolation-level "Default isolation level for txns" :serializable)
+
+(def ysql-port 5433)
+
+(defn db-spec
+  "Assemble a JDBC connection specification for a given Jepsen node."
+  [node]
+  {:dbtype         "postgresql"
+   :dbname         "postgres"
+   :classname      "org.postgresql.Driver"
+   :host           (name node)
+   :port           ysql-port
+   :user           "postgres"
+   :password       ""
+   :loginTimeout   (/ max-timeout 1000)
+   :connectTimeout (/ max-timeout 1000)
+   :socketTimeout  (/ max-timeout 1000)})
+
+(defn close-conn
+  "Given a JDBC connection, closes it and returns the underlying spec."
+  [conn]
+  (when-let [c (j/db-find-connection conn)]
+    (.close c))
+  (dissoc conn :connection))
+
+(defn conn-wrapper
+  "Constructs a network client for a node, and opens it"
+  [node]
+  (rc/open!
+    (rc/wrapper
+      {:name  node
+       :open  (fn open []
+                (util/timeout max-timeout
+                              (throw (RuntimeException.
+                                       (str "Connection to " node " timed out")))
+                              (util/retry 0.1
+                                          (let [spec  (db-spec node)
+                                                conn  (j/get-connection spec)
+                                                spec' (j/add-connection spec conn)]
+                                            (assert spec')
+                                            spec'))))
+       :close close-conn
+       :log?  true})))
+
+(defn exception->op
+  "Takes an exception and maps it to a partial op, like {:type :info, :error
+  ...}. nil if unrecognized."
+  [e]
+  (when-let [m (.getMessage e)]
+    (condp instance? e
+      java.sql.SQLTransactionRollbackException
+      {:type :fail, :error [:rollback m]}
+
+      java.sql.BatchUpdateException
+      (if (re-find #"getNextExc" m)
+        ; Wrap underlying exception error with [:batch ...]
+        (when-let [op (exception->op (.getNextException e))]
+          (update op :error (partial vector :batch)))
+        {:type :info, :error [:batch-update m]})
+
+      org.postgresql.util.PSQLException
+      (condp re-find (.getMessage e)
+        #"(?i)Conflicts with [a-z]+ transaction"
+        {:type :fail, :error [:conflicting-transaction m]}
+
+        #"(?i)Catalog Version Mismatch"
+        {:type :fail, :error [:catalog-version-mismatch m]}
+
+        #"(?i)Operation expired"
+        {:type :fail, :error [:operation-expired m]}
+
+        {:type :info, :error [:psql-exception m]})
+
+      clojure.lang.ExceptionInfo
```

Comment:
In which cases this exception is thrown?

Reply:
Please add a comment to the code, so it will be clear for whoever reading that code.